exasol/ai/mcp/server/main.py

```diff
@@ -323,11 +258,27 @@ def get_connection_factory(
         "compression": True,
     }
     conn_kwargs.update(extra_kwargs)
-    if auth_method == AuthenticationMethod.PASSWORD:
-        conn_kwargs["password"] = env.get(ENV_PASSWORD)
+
+    # Infer the authentication method. If neither password nor bearer token are
+    # provided assume that the MCP server is protected with OpenID Connect and the
+    # access token can be found in the MCP server context. This token will be used
+    # to connect to the database. The database, in its turn, will have to verify the
+    # token with the same identity provider as used by the MCP server.
+    # Otherwise, use the password or a bearer token, whichever is provided. The MCP
+    # server may still be protected with the OpenID, but in this case it is irrelevant
+    # as far as the database connection is concerned.
+    use_open_id = False
+    if ENV_PASSWORD in env:
+        conn_kwargs["password"] = env[ENV_PASSWORD]
+    elif ENV_ACCESS_TOKEN in env:
+        conn_kwargs["access_token"] = env[ENV_ACCESS_TOKEN]
+    elif ENV_REFRESH_TOKEN in env:
+        conn_kwargs["refresh_token"] = env[ENV_REFRESH_TOKEN]
+    else:
+        use_open_id = True
 
     def connection_factory() -> ExaConnection:
-        if auth_method == AuthenticationMethod.OPEN_ID:
+        if use_open_id:
```

Comment:
If we use open_id, we actually need to recreate a new connection object when ever the server is called, because each call could be a different user. This means we can't use the connection caching as before. We could only use caching if we have a user context or something like that.

Reply:
It is valid. The problem is that all MCP server users must have the same permissions. Effectively I can authorize someone else to do something on the DB, without knowing this. This is very limiting, and perhaps will sound silly if you describe it to a customer. We need to get away from this single user model. But this is a subject for another epic I guess, as it's not directly related to OpenID support.